### PR TITLE
[POAE7-2434] Migrate substrait-cpp of function lookup

### PR DESCRIPTION
**Pull request by haiweizh**
### What changes were proposed in this pull request?

Migrate substrait-cpp of function lookup, use [substrait-cpp].

### Why are the changes needed?

Migrate substrait-cpp of function lookup.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Uts.

### Which label does this PR belong to?

CODEREFACTOR,cider,test

### CODE REVIEW

**Comment by haiweizh** [comment]
old pr is closed unexpectedly. https://github.com/intel/BDTK/pull/106
Use ExternalProject to input and use substrait-cpp.

**Review by winningsix — DISMISSED** [review]
Good refactor. Please take a care of failed checks.

**Comment by yma11** [comment]
@haiweizh There is still a building error, please resolve it.

File: .gitmodules [hunk 1]
```diff
@@ -22,3 +22,6 @@
 [submodule "thirdparty/yaml-cpp"]
 	path = thirdparty/yaml-cpp
 	url = https://github.com/jbeder/yaml-cpp.git
+[submodule "thirdparty/substrait-cpp"]
```
Comment (xieqi) [inline .gitmodules:25]: why still need submodule for substrait-cpp, you can git clone the code with externalproject

File: cider/function/FunctionLookupEngine.cpp [hunk 2]
```diff
@@ -73,12 +67,11 @@ const SQLOps FunctionLookupEngine::getFunctionScalarOp(
...
   const std::string& func_name = function_signature.func_name;
-  const auto& functionSignature =
-      cider::function::substrait::SubstraitFunctionSignature::of(
-          func_name, function_signature.arguments, function_signature.return_type);
-  const auto& functionOption =
-      scalar_function_look_up_ptr_->lookupFunction(functionSignature);
-  if (functionOption.has_value()) {
+  const std::vector<io::substrait::TypePtr>& arguments = function_signature.arguments;
+  const io::substrait::TypePtr& return_type = function_signature.return_type;
+  const auto& functionVariantPtr =
```
Comment (xieqi) [inline cider/function/FunctionLookupEngine.cpp:72]: functionVariantPtr --> function_variant_ptr and other local variables to keep variable naming consistent

File: CMakeLists.txt [hunk 3]
```diff
@@ -122,6 +122,9 @@ add_definitions(-DRAPIDJSON_HAS_STDSTRING)
 # robin-hood-hashing
 include_directories(thirdparty/robin-hood-hashing/src/include)
 
+set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake;${CMAKE_MODULE_PATH}")
```
[the remark this answers is not in the document]
Comment (haiweizh) [reply CMakeLists.txt]: Done.

File: .gitmodules (hunk 1 above)
Comment (haiweizh) [reply .gitmodules]: Done.

File: cider/function/FunctionLookupEngine.cpp (hunk 2 above)
Comment (haiweizh) [reply cider/function/FunctionLookupEngine.cpp]: Done.

**Comment by haiweizh** [comment]
PR ready. Plz help to take a review. @winningsix @xieqi @yma11 @chaojun-zhang

**Comment by haiweizh** [comment]
PR ready. Plz help to take a review. @winningsix @xieqi @yma11 @jikunshang @chaojun-zhang

File: cider/tests/CiderCompileModuleTest.cpp [hunk 4]
```diff
@@ -32,17 +32,43 @@ std::string getDataFilesPath() {
...
   std::stringstream buffer;
   buffer << sub_json.rdbuf();
   std::string sub_data = buffer.str();
-  substrait::Plan sub_plan;
+  ::substrait::Plan sub_plan;
   google::protobuf::util::JsonStringToMessage(sub_data, &sub_plan);
-
   auto ciderCompileModule =
       CiderCompileModule::Make(std::make_shared<CiderDefaultAllocator>());
+  ciderCompileModule->feedBuildTable(std::move(buildCiderBatch()));
```
Comment (jikunshang) [inline cider/tests/CiderCompileModuleTest.cpp:71]: if this is not a join query, why need this?

File: cider/function/FunctionLookupEngine.cpp [hunk 5]
```diff
@@ -22,39 +22,33 @@
...
-  scalar_function_look_up_ptr_ =                                                      \
-      std::make_shared<cider::function::substrait::SubstraitScalarFunctionLookup>(    \
-          cider_internal_function_ptr, func_mappings);                                \
-  aggregate_function_look_up_ptr_ =                                                   \
-      std::make_shared<cider::function::substrait::SubstraitAggregateFunctionLookup>( \
-          cider_internal_function_ptr, func_mappings);                                \
-  extension_function_look_up_ptr_ =                                                   \
-      std::make_shared<cider::function::substrait::SubstraitScalarFunctionLookup>(    \
-          extension_function_ptr, func_mappings);
+#define LOAD_EXTENSION_YAML_AND_INITIALIZE_FUNCTION_LOOK_UP(                            \
```
Comment (jikunshang) [inline cider/function/FunctionLookupEngine.cpp:25]: I think it's better to convert this macro to a method.

File: cider/function/FunctionLookupEngine.cpp [hunk 6]
```diff
@@ -96,12 +89,11 @@ const SQLAgg FunctionLookupEngine::getFunctionAggOp(
...
-      cider::function::substrait::SubstraitFunctionSignature::of(
-          func_name, function_signature.arguments, function_signature.return_type);
-  const auto& functionOption =
-      aggregate_function_look_up_ptr_->lookupFunction(functionSignature);
-  if (functionOption.has_value()) {
+  const std::vector<io::substrait::TypePtr>& arguments = function_signature.arguments;
+  const io::substrait::TypePtr& return_type = function_signature.return_type;
+  const auto& function_variant_ptr = aggregate_function_look_up_ptr_->lookupFunction(
+      {func_name, arguments, return_type});
+  if (function_variant_ptr != nullptr) {
```
Comment (jikunshang) [inline cider/function/FunctionLookupEngine.cpp:96]: `if (function_variant_ptr)`

Comment (haiweizh) [reply cider/function/FunctionLookupEngine.cpp]: Done.

File: cider/tests/CiderCompileModuleTest.cpp (hunk 4 above)
Comment (haiweizh) [reply cider/tests/CiderCompileModuleTest.cpp:71]: Will coredump.

File: cider/function/FunctionLookupEngine.cpp (hunk 5 above)
Comment (haiweizh) [reply cider/function/FunctionLookupEngine.cpp:25]: Use different function mapping type, macro is better.

Comment (jikunshang) [reply cider/function/FunctionLookupEngine.cpp:25]: how about template.

**Review by winningsix — APPROVED** [review]

File: cider/tests/CiderCompileModuleTest.cpp (hunk 4 above)
Comment (yma11) [reply cider/tests/CiderCompileModuleTest.cpp:71]: +1.

Comment (haiweizh) [reply cider/tests/CiderCompileModuleTest.cpp:71]: CiderCompileModuleTest will coredump if not feedBuildTable after this code refactor. I think maybe this is a previous bug？@yma11

File: cider/function/FunctionLookupEngine.cpp (hunk 5 above)
Comment (haiweizh) [reply cider/function/FunctionLookupEngine.cpp]: Done.

**Review by winningsix — APPROVED** [review]

File: cider/tests/CiderCompileModuleTest.cpp (hunk 4 above)
Comment (yma11) [reply cider/tests/CiderCompileModuleTest.cpp:71]: I think we need take a look at it. I think you code refactor has no effect on compile module.

Comment (haiweizh) [reply cider/tests/CiderCompileModuleTest.cpp:71]: Will verify this tast case in this subtask. [POAE7-2597] @yma11

**Review by yma11 — APPROVED** [review]

**Pull request merged**